modules/standard/Sets.chpl

```diff
@@ -0,0 +1,700 @@
+    */
+    proc init(type eltType, param parSafe=false) {
+      this.eltType = eltType;
+      this.parSafe = parSafe;
+    }
+
+    /*
+      Initialize this set with a unique copy of each element contained in
+      `iterable`. If an element from `iterable` is already contained in this
+      set, it will not be added again. The formal `iterable` must be a type
+      with an iterator defined for it.
+
+      :arg iterable: A collection of elements to add to this set.
+      :arg parSafe: If `true`, this set will use parallel safe operations.
+    */
+    proc init(type eltType, param parSafe, iterable) {
+      this.eltType = eltType;
+      this.parSafe = parSafe;
+      this.complete();
+      //
+      // TODO: Is there a reliable way to get iterable size, here? If so, we
+      // could make a call to `requestCapacity`.
+      //
+      for x in iterable do
+        _dom.add(x);
+    }
+
+    /*
+      Initialize this set with a copy of each of the elements contained in
+      the set `other`. This set will inherit the `parSafe` value of the
+      set `other`.
+
+      :arg other: A set to initialize this set with.
+    */
+    proc init=(const ref other: set(?t, ?)) {
+      this.eltType = t;
+      this.parSafe = other.parSafe;
+      this.complete();
+      for x in other {
+        //
+        // TODO: Not entirely sure if this inner copy is necessary, but we
+        // want to avoid potentially locking over and over.
+        //
+        var cpy = x;
+        _dom.add(cpy);
+      }
+    }
+
+    pragma "no doc"
+    inline proc deinit() {
+      clear();
+    }
+
+    pragma "no doc"
+    inline proc _enter() {
+      if parSafe then
+        _lock$.lock();
+    }
+
+    pragma "no doc"
+    inline proc _leave() {
+      if parSafe then
+        _lock$.unlock();
+    }
+ 
+    /*
+      Add a copy of the element `x` to this set. Does nothing if this set
+      already contains an element equal to the value of `x`.
+
+      :arg x: The element to add to this set.
+    */
+    proc add(in x: eltType) {
+      _enter();
+      const dbg = _dom.add(x);
+      _leave();
+    }
+
+    /*
+      Returns `true` if the given element is a member of this set, and `false`
+      otherwise.
+
+      :arg x: The element to test for membership.
+      :return: Whether or not the given element is a member of this set.
+      :rtype: `bool`
+    */
+    proc const contains(const ref x: eltType): bool {
+      _enter();
+      var result = _dom.contains(x);
+      _leave();
+      return result;
+    }
+
+    /*
+      Returns `true` if this set shares no elements in common with the set
+      `other`, and `false` otherwise.
+
+      :arg other: The set to compare against.
+      :return: Whether or not this set and `other` are disjoint.
+      :rtype: `bool`
+    */
+    proc const isDisjoint(const ref other: set(eltType, ?)): bool {
+      _enter();
+
+      var result = true;
+      for x in other do
```

Comment:
Serial iteration does not seem to acquire a lock on `other`; for consistency shouldn't you do `other._enter()` and `other._leave()`? If the answer is no, you're risking data races since you enter a critical section in one but not the other; if the answer is yes, you risk deadlock if `set1.isDisjoint(set2)` and `set2.isDisjoint(set1)` occur concurrently. Perhaps you should create some kind of locking order based on `_lock$` wide pointer and address? What do you think @mppf

Reply:
https://github.com/pnnl/chgl/blob/4c1dd6f050d4f1d7afed2f7302e020f5eea6b429/src/Utilities.chpl#L36-L106

How I handle it, just as a guide of reference.

Reply:
How would I enforce a lock ordering if the `set` acquires a lock in its serial iterator? 

I'm not even sure if iterators should be threadsafe. The reason is that even if an iterator is threadsafe there is always the possibility of reference invalidation if another thread goes and adds/removes an element to the container.

Do you think it would just be better to let the serial iterator not be threadsafe but then enforce a lock ordering before iterating in `isDisjoint`?

The bigger problem conceptually here for me is that both sets might not have `parSafe==true`, basically I'm kind of at a loss for what the granularity of the critical sections should be here.

Since either set can have parallel safety on _or_ off, I would probably only want to acquire the locks in a fixed order if both sets have parallel safety `true`. 

Think you can help me muddle through this madness, @LouisJenkinsCS?

Reply:
When it comes to acquiring in a global locking order where one set is `parSafe` and the other is not, you add a conditional for this case and handle it accordingly. This might not be necessary since `_enter` and `_leave` already result in a NOP if `parSafe` is false. Hence this should suffice for current implementation for `isDisjoint`. Granularity should be for the entire operation since the the entire operation is handled internally. 

I'd personally ditch associative domains (no offense) and implement some kind of transaction-based set. Maybe something like this: https://www.google.com/amp/s/blog.acolyer.org/2015/10/27/read-log-update-a-lightweight-synchronization-mechanism-for-concurrent-programming/amp/

Read-Log-Update could enable concurrent iteration with modifications too. You'd just need to track per-task write logs. I'm not saying it's easy, there is a reason why non-blocking data structures are a thing; locks have numerous disadvantages and complexities that cannot be understated.

Reply:
Thanks Louis, I appreciate the tips. WRT using a different implementation, I think @bradcray is really intent on dogfooding here, hence the use of the associative domain.

Reply:
I don't think having specialized data structures is against the whole "eating your own dog food" philosophy; I'd say that if you implemented something new (novel even) that worked, written entirely in Chapel with constructs that are available to the user (I.E no specialized runtime components), then you are still upholding to this philosophy.

Personally, I prefer giving the user the option to control granularity of locks by making `parSafe` an argument to each function, allowing a smarter user to avoid a lot of deadlock conditions. Then document that deadlock is an issue for iteration if you acquire the lock and then try to call some other function with `parSafe=true` and that while iterating they may safely call functions with `parSafe=false`.

Reply:
Historically for the array implementations, the spec says that parallel modification during iteration leads to undefined behavior. We could do the same thing here.

Reply:
Seriously though, making a optional `parSafe` method argument can work in some cases.